File: 灵茶/图论/并查集/0547.py
class UnionFind:
    def __init__(self):
        self.father = {}
        self.size = 0
    
    def find(self, x):
        root = x

        while self.father[root] != None:
            root = self.father[root]
        
        while root != x:
            original_father = self.father[x]
            self.father[x] = root
            x = original_father

        return root
    
    def merge( self, x, y):
        root_x, root_y = self.find(x), self.find(y)

        if root_x != root_y:
            self.father[root_x] = root_y
            self.size -= 1
        
    def is_connected(self, x, y):
        return self.find(x) == self.find(y)
    
    def add(self, x):
        if x not in self.father:
            self.father[x] = None
            self.size += 1

class Solution:
    def findCircleNum(self, M: List[List[int]]) -> int:
        uf = UnionFind()
        for i in range(len(M)):
            uf.add(i)
            for j in range(i):
                if M[i][j]:
                    uf.merge(i, j)
        
        return uf.size

File: 灵茶/图论/并查集/0547.java
import java.util.*;

class UnionFind {

    private Map<Integer, Integer> father;

    private int size;

    public UnionFind() {
        father = new HashMap<Integer, Integer>();
        size = 0;
    }

    public void add(int x) {
        if (!father.containsKey(x)) {
            father.put(x, null);
            size++;
        }
    }

    public void merge(int x, int y) {
        int rootX = find(x);
        int rootY = find(y);

        if (rootX != rootY) {
            father.put(rootX, rootY);
            size--;
        }
    }

    public int find(int x) {
        int root = x;
        while (father.get(root) != null) {
            root = father.get(root);
        }

        while (x != root) {
            int original_father = father.get(x);
            father.put(x, root);
            x = original_father;
        }

        return root;
    }

    public boolean isConnected(int x, int y) {
        return find(x) == find(y);
    }

    public int getSize() {
        return size;
    }

}

class Solution {
    public int findCircleNum(int[][] isConnected) {
        UnionFind UF = new UnionFind();
        for (int i = 0; i < isConnected.length; i++) {
            UF.add(i);
            for (int j = 0; j < i; j++) {
                if (isConnected[i][j] == 1) {
                    UF.merge(i, j);
                }
            }
        }

        return UF.getSize();
    }
}
